packages/models/src/role/types.ts

```diff
@@ -1,11 +1,108 @@
+import { TYPE_LESSON } from '../lesson';
+import { TYPE_ACTIVITY } from '../activity';
 
-export interface Role<T> {
-  user: User;
-  role: T;
-}
+const SubjectViewerPrivileges = ['view', 'clone', 'import'];
+const SubjectEditorPrivileges = [...SubjectViewerPrivileges, 'updateContent'];
+const SubjectOwnerPrivileges = [
+  ...SubjectEditorPrivileges,
+  'editPermissions',
+  'deleteSubject',
+];
+const ActivityParticipatorPrivileges = ['participate'];
+const ActivityModeratorPrivileges = [
+  ...ActivityParticipatorPrivileges,
+  'editSubject',
+  'editActivity',
+  'editParticipants',
+];
+const ActivityOwnerPermissions = [
+  ...ActivityModeratorPrivileges,
+  'editPermissions',
+  'deleteActivity',
+  'cloneActivity',
+  'exportSubjects',
+];
 
-export interface NormalizedRole<T> {
-  user: User['id'];
-  role: T;
-}
+const UserGroupMembersPermissions = [
+  'exitGroup', // do we need such permission?
+  // 'inheritsPermissionsOfTheUserGroup' <-- goes without explicit check but needs different implementation todo
+];
+const UserGroupViewerPermissions = [
+  ...UserGroupMembersPermissions,
+  'viewGroupMembers',
+];
+const UserGroupAdminPermissions = [
+  ...UserGroupViewerPermissions,
+  'editNonAdminGroupMembers',
+  'canAddMembersByInvite',
+  'canEditMembers',
+  'canEditViewers',
+];
+const UserGroupOwnerPermissions = [
+  ...UserGroupAdminPermissions,
+  'deleteGroup',
+  'editAdmins',
+  'editOwners',
+];
+
+type RolePrivileges = {
+  [category: string]: {
+    [role: string]: string[];
+  };
+};
+export const rolePrivileges: RolePrivileges = {
+  Subject: {
+    Viewer: SubjectViewerPrivileges,
+    Editor: SubjectEditorPrivileges,
+    Owner: SubjectOwnerPrivileges,
+  },
+  Activity: {
+    Participator: ActivityParticipatorPrivileges,
+    Moderator: ActivityModeratorPrivileges,
+    Owner: ActivityOwnerPermissions,
+  },
+  UserGroup: {
+    Member: UserGroupMembersPermissions,
+    Viewer: UserGroupViewerPermissions,
+    Admin: UserGroupAdminPermissions,
+    Owner: UserGroupOwnerPermissions, // we might need rule a that we can't leave group owner-less (or cover it with business logic rather)
+  },
+};
+
+const getRoleDefinitionsByObjectType = (objectType: String): any => {
+  switch (objectType) {
+    case TYPE_STEP:
+    case TYPE_CHAPTER:
+    case TYPE_LESSON:
+      return rolePrivileges.Subject;
+    case TYPE_ACTIVITY:
+      return rolePrivileges.Activity;
+    // case TYPE_USER_GROUP:
+    //   return rolePrivileges.UserGroup;
+    default:
+      throw new Error(
+        'Unable to find RolePrivileges definition of object type for ' +
+          objectType,
+      );
+  }
+};
+
+export const getRolesWithAction = (
+  objectType: string,
+  action: string,
+): string[] => {
+  const roleDefinitions = getRoleDefinitionsByObjectType(objectType);
+  return Object.keys(roleDefinitions).filter((role: string) =>
+    roleDefinitions[role].includes(action),
+  );
+};
+export const hasPermissions = (
```

Comment:
I'd name this `isRoleAction` or something like that. `hasPermission` looks like end product, a function that will actually resolve holder permission.